src/benchmarklib/table_builder.hpp

```diff
@@ -0,0 +1,240 @@
+  }
+}
+
+}  // namespace
+
+namespace opossum {
+
+/**
+ * Helper to build a table with a static column layout, specified by ctor args types and names. Keeps a
+ * data vector for each column and appends values to them in append_row(). For nullable columns an additional
+ * is_null_vector is kept. Automatically creates chunks in accordance with the specified chunk size.
+ */
+template <typename... DataTypes>
+class TableBuilder {
+ public:
+  // names is a list of strings defining column names, types is a list of equal length defining respective column types
+  // types may contain std::optional<?>, which will result in a nullable column, otherwise columns are not nullable
+  template <typename... Names>
+  TableBuilder(size_t chunk_size, const hana::tuple<DataTypes...>& types, const hana::tuple<Names...>& names,
+               opossum::UseMvcc use_mvcc, size_t estimated_rows = 0)
+      : _use_mvcc(use_mvcc), _estimated_rows_per_chunk(std::min(estimated_rows, chunk_size)) {
+    BOOST_HANA_CONSTANT_ASSERT(hana::size(names) == hana::size(types));
+
+    // Iterate over the column types/names and create the columns.
+    auto column_definitions = TableColumnDefinitions{};
+    hana::fold_left(hana::zip(names, types), column_definitions,
+                    [](auto& definitions, const auto& name_and_type) -> decltype(definitions) {
+                      auto name = name_and_type[hana::llong_c<0>];
+                      auto type = name_and_type[hana::llong_c<1>];
+
+                      auto data_type = data_type_from_type<get_value_type<decltype(type)>>();
+                      auto is_nullable = is_optional<decltype(type)>();
+
+                      definitions.emplace_back(name, data_type, is_nullable);
+
+                      return definitions;
+                    });
+    _table = std::make_shared<Table>(column_definitions, TableType::Data, chunk_size, use_mvcc);
+
+    // Reserve some space in the vectors
+    hana::for_each(_data_vectors, [&](auto&& vector) { vector.reserve(_estimated_rows_per_chunk); });
+    hana::for_each(_is_null_vectors, [&](auto&& is_null_vector) {
+      if constexpr (std::decay_t<decltype(is_null_vector)>::has_value) {
+        is_null_vector.value().reserve(_estimated_rows_per_chunk);
+      }
+    });
+  }
+
+  std::shared_ptr<Table> finish_table() {
+    if (_current_chunk_row_count() > 0) {
+      _emit_chunk();
+    }
+
+    return _table;
+  }
+
+  // Types == DataTypes, we deduce again to get universal references, so we can call with lvalues
+  template <typename... Types>
+  void append_row(Types&&... values) {
+    auto values_tuple = hana::make_tuple(std::forward<decltype(values)>(values)...);
+    BOOST_HANA_CONSTANT_ASSERT(hana::size(hana::tuple<DataTypes...>()) == hana::size(values_tuple));
+
+    // Create tuples ([&data_vector0, &is_null_vector0, value0], [&data_vector1, &is_null_vector1, value1], ...)
+    auto data_vectors_and_is_null_vectors_and_values = hana::zip_with(
```

Comment:
why do you zip here and then do "hana::for_each" below in a different step? Why don't you do emplace_backs()s right here already?

Reply:
As a reader I would probably not expect anything to happen in a zip_with lambda apart from zipping, also I'm pretty sure I just stuck to that pattern because it was already there when I started working on the table_builder. You're right though, it's easier to do it in one step, I will change it.

Reply:
ah, maybe zip_with does not guarantee order of execution, where fold_left does? in that case we should probably use two steps :(

Reply:
I found a comment in the ["old" table builder](https://github.com/hyrise/hyrise/blob/master/src/benchmarklib/tpch/tpch_table_generator.cpp): `Create a tuple ((column_name0, column_type0), (column_name1, column_type1), ...) so we can iterate over the columns. fold_left as below does this in order, I think boost::hana::zip_with() doesn't, which is why I'm doing two steps here.` However, I can't find that in the boost hana documentation. Do you have a link where I can read up on boost hana's order of execution?

Reply:
In this case the order of execution does not matter, so I would still do it in one step and add a comment about order of execution.

Reply:
I think "Side effects and purity" [here](https://www.boost.org/doc/libs/1_65_0/libs/hana/doc/html/index.html) talks about this. "Except where specified otherwise, all functions used with Hana (i.e. used in higher order algorithms) should be pure. In particular, functions passed to higher order algorithms are not guaranteed to be called any specific number of times"

So I have to back away from my initial proposal and opt for staying with zip_with followed by fold_left

Reply:
Alright, thanks a lot! It also says "However, some exceptional algorithms like for_each do expect impure functions, and they guarantee an order of evaluation." I prefer for_each over fold_left, to avoid that awkward duality we discussed before.